lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -414,16 +414,40 @@ SPIRVType *LLVMToSPIRV::transType(Type *T) {
     if (Name == getSPIRVTypeName(kSPIRVTypeName::ConstantPipeStorage))
       return transType(getPipeStorageType(M));
 
-    auto *Struct = BM->openStructType(T->getStructNumElements(), Name.str());
+    const size_t MaxNumElements = 65535 - 2;
+    const size_t NumElements = T->getStructNumElements();
+    size_t SPIRVStructNumElements = NumElements;
+    if (NumElements > MaxNumElements &&
+        BM->isAllowedToUseExtension(
+            ExtensionID::SPV_INTEL_long_constant_composite)) {
+      SPIRVStructNumElements = MaxNumElements;
+    }
+
+    auto *Struct = BM->openStructType(SPIRVStructNumElements, Name.str());
     mapType(T, Struct);
 
+    if (NumElements > MaxNumElements &&
+        BM->isAllowedToUseExtension(
+            ExtensionID::SPV_INTEL_long_constant_composite)) {
+      uint64_t NumOfContinuedInstructions = NumElements / MaxNumElements - 1;
```

Comment:
If `NumElements == MaxNumElements + 1`, then the result of division would be 1, which means that we need zero continued instructions and it doesn't seem to be correct. Am I missing something?

Reply:
In the current version, this last one element will be handled on lines 438-443, where I calculate modulo. I didn't look closely through Mikhail's suggestion, but I believe it will be handled there as well.